### ISSUE TITLE
[3.1.8] Main class is not found when it does not extend GrailsAutoConfiguration itself, even if a superclass does.

### ISSUE THREAD

**Issue by budjb** [grails/grails-core#10009]
### Expected Behaviour

I have a scenario where it makes sense to create a superclass to an application's main application class. I have created a superclass that extends `GrailsAutoConfiguration`, as the default main application does, and added some logic that the main application class will inherit. I expect that the `bootRun` task will successfully find the application's main class with this object structure:

```
GrailsAutoConfiguration
|---ApplicationSuperClass
    |---Application
        |---static main(String[] args)
```

### Actual Behaviour
The `bootRun` task does not successfully find the application's main class. The only way the `bootRun` task finds the main class successfully is when it directly extends `GrailsAutoConfiguration`.

### Example Application

- TODO: link to github repository with example that reproduces the issue

**Comment by graemerocher** [comment]
Yeah this is hard to fix because we use ASM to figure out the `Application` class and it only allows access to the super class name, not the whole class hierarchy: https://github.com/grails/grails-core/blob/master/grails-bootstrap/src/main/groovy/org/grails/io/support/MainClassFinder.groovy#L150

**Comment by budjb** [comment]
@graemerocher Ah I see. There's also a misleading error the first time that example application is run after a clean with a `NullPointerException`. Subsequent attempts without a clean produce the proper message stating that the main class could not be found.  Perhaps that issue can be resolved?

**Comment by graemerocher** [comment]
Certainly
